modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java

```diff
@@ -2526,36 +2526,60 @@ private Discriminator recursiveGetDiscriminator(Schema sc, OpenAPI openAPI) {
             if (composedSchema.getOneOf() != null && composedSchema.getOneOf().size() != 0) {
                 // All oneOf definitions must contain the discriminator
                 Integer hasDiscriminatorCnt = 0;
+                Integer hasNullTypeCnt = 0;
                 Set<String> discriminatorsPropNames = new HashSet<>();
                 for  (Schema oneOf: composedSchema.getOneOf()) {
+                    if (ModelUtils.isNullType(oneOf)) {
+                        // The null type does not have a discriminator. Skip.
+                        hasNullTypeCnt++;
+                        continue;
+                    }
```

Comment:
@sebastien-rosset @wing328 @jimschubert what should we allow here, valid objects and empty state values? 
Null only? Null and empty dict?
Is this a language specific decision? If so we should separate it off as a separate method.
- For example for python values like empty dict {} and None are valid empty values here.
- For golang the zero value of a map is nil.

Reply:
The "null type" (supported in OAS >=3.1) is when the OAS schema is:
```yaml
type: 'null'
```

It means the value in the payload must be null. This can typically be list as one child of "oneOf".

Reply:
So I am not asking if it is in the specification. I understand that it is. I am asking if we should also allow an empty object payload here as synonymous with null because that is a python zeto state like null/None.

Reply:
The change is to specifically cover the following scenario:
```yaml
oneOf:
  - type: null
  - $ref: abc   # possibly a list of inline types or references to types.
```
Without this PR, the count of oneOf children was not be equal to the number of discriminators. That's because the "null" value does not have a discriminator. As a result, the disc.setPropertyName() was not set, which is incorrect.

> I am asking if we should also allow an empty object payload

Can you give how this would be modeled in the OAS document? Here we are parsing OAS schemas, not payloads.

Reply:
We are describing payloads here. An empty object represents an rmpty dict. Should this definition be valid?
```
ComposedSchema:
OneOf
- type: null
- type: object
- AnyType
- Pet
discriminator:
  PropertyName: petType

AnyType:
  description: I allow any type including null
```

Reply:
I think that AnyType should not be allowed because list and other primitives are different data types than an intended dict/pojo.
My preference is to allow null and empty object but log a warning to the user that this definition is not complaint with the v3 specification bit we allow it.

Reply:
Thanks for the clarification. Here is one statement in the OAS 3.1 spec:

> When using the discriminator, inline schemas will not be considered.

The way I interpret is that if there is a mix of oneOf children, some through a $ref, others inline, then at runtime:
1. For the inline schemas, use JSON schema validation to determine if the payload matches the schema. Don't use the discriminator.
1. For $ref schemas, don't use JSON schema validation, and instead use the discriminator.

So if my understand is correct, the schema you have provided is a valid OAS schema. HOWEVER, at runtime it will fail validation in many situations, because of the presence of "AnyType". It will fail when the payload is "null", object and Pet. It would work if the payload is a primitive type or an array. So it would be much better to write the OAS spec differently.

I'd rather address these use cases in separate PRs. IMO, we would ideally add validation rules to the `--validate` option.

In the OAS 3.0.3 spec:

> When used, the discriminator will be the name of the property that decides which schema definition validates the structure of the model. As such, the discriminator field MUST be a required field.

Reply:
Our users could just as easily write a spec that looks like this:
```
Animal:
oneOf:
- $ref: NullType
- $ref: Pet
discriminator:
  propertyName: animalType

NullType:
type: "null"
```
So I am okay addressing the object type question in the future.
You are proposing accepting nulls in a payload when a strict interpretation of the spec would require the presence of a pojo with a discriminator.propertyName at not null. So I feel like we're in uncharted territory here.

You mention that it would fail because it would double match null on AnyType and type: None. That wasn't the point that I was trying to make. The questions that I am trying to get your answers on are:
- Should AnyType be allowed in oneOf generally?
- Should empty object `type: object` be allowed in oneOf?
I think that we should allow `type: "null"` in oneOf but I do not think that that behavior is defined in the draft 3.1.0 spec yet.

Reply:
When mixing types in oneOf, should we disallow the use of allOf + anyOf in the composed schema.
Otherwise you could have:
```
Creature:
allOf:
- $ref: Animal
oneOf:
- type: int
- $ref: Pet
```

Reply:
> Our users could just as easily write a spec that looks like this:
> 
> ```
> Animal:
> oneOf:
> - $ref: NullType
> - $ref: Pet
> discriminator:
>   propertyName: animalType
> 
> NullType:
> type: "null"
> ```
> 
> Should we handle that referenced null case? I think that we should.
> 

I don't think this matches what is written in the OAS 3.1 spec. If the type is a ref and there is a discriminator, the discriminator must be used. That's my understand of the spec. So your YAML is not equivalent.

> You are proposing accepting nulls in a payload when a strict interpretation of the spec would require the presence of a pojo with a discriminator.propertyName at not null. So I feel like we're in uncharted territory here.

Which part of the spec (3.1) makes you think that? It is indeed my understanding, reading the many threads over the null type that the OAS spec authors intend to replace "nullable" with `type: 'null'` as oneOf child.

> 
> In your comment you mention that it would fail because it would double match null on AnyType and type: None. That wasn't the point that I was trying to make. The questions that I am trying to get your answers on are:
> 
> * Should AnyType be allowed in oneOf generally?
>   

Yes, absolutely, for example below it's a sensible way to say the payload can be anything, including null. I think there is even one thread about this specific item in the draft discussions that led to 3.1, I will try to dig it up.

```yaml
oneOf:
- type: 'null'
- description: any type
```

>   * My vote is no. If they have this in the oneOf then it has to be the only entry. Also then they can't have any allOf definition. Which is equivalent to just using the AnyType directly.
> * Should empty object `type: object` be allowed in oneOf?
> 
> I think that we should allow `type: "null"` in oneOf with a discriminator but I do not think that that behavior is defined in the draft 3.1.0 spec yet.

Reply:
> When mixing types in oneOf, should we disallow the use of allOf + anyOf in the composed schema?

Sorry I'm not getting this part. Did you mean a schema like this where there is "anyOf"? In the example you provide it is allOf + oneOf, not anyOf.

```
Creature:
allOf:
- $ref: Animal
anyOf:
- type: int
- $ref: Pet
```

Also, can you help me understand how your questions are relevant to this PR? I'm just trying to handle the "null" case here, I'm certainly not claiming to cover all possible composed scenarios + "null". Is there something specific you see that I am breaking with the support of the "null" value? Or are you extending the discussion to other use cases?

> Otherwise you could have the invalid:
> 
> ```
> Creature:
> allOf:
> - $ref: Animal
> oneOf:
> - type: int
> - $ref: Pet
> ```

Reply:
This is more warranted as a separate discussion.

Reply:
> > Our users could just as easily write a spec that looks like this:
> > ```
> > Animal:
> > oneOf:
> > - $ref: NullType
> > - $ref: Pet
> > discriminator:
> >   propertyName: animalType
> > 
> > NullType:
> > type: "null"
> > ```
> > 
> > 
> > Should we handle that referenced null case? I think that we should.
> 
> I don't think this matches what is written in the OAS 3.1 spec. If the type is a ref and there is a discriminator, the discriminator must be used. That's my understand of the spec. So your YAML is not equivalent.
> 
> > You are proposing accepting nulls in a payload when a strict interpretation of the spec would require the presence of a pojo with a discriminator.propertyName at not null. So I feel like we're in uncharted territory here.
> 
> Which part of the spec (3.1) makes you think that? It is indeed my understanding, reading the many threads over the null type that the OAS spec authors intend to replace "nullable" with `type: 'null'` as oneOf child.
> 
> > In your comment you mention that it would fail because it would double match null on AnyType and type: None. That wasn't the point that I was trying to make. The questions that I am trying to get your answers on are:
> > 
> > * Should AnyType be allowed in oneOf generally?
> 
> Yes, absolutely, for example below it's a sensible way to say the payload can be anything, including null. I think there is even one thread about this specific item in the draft discussions that led to 3.1, I will try to dig it up.
> 
> ```yaml
> oneOf:
> - type: 'null'
> - description: any type
> ```
> 
> > * My vote is no. If they have this in the oneOf then it has to be the only entry. Also then they can't have any allOf definition. Which is equivalent to just using the AnyType directly.
> > * Should empty object `type: object` be allowed in oneOf?
> > 
> > I think that we should allow `type: "null"` in oneOf with a discriminator but I do not think that that behavior is defined in the draft 3.1.0 spec yet.
> 
> It is, though admittedly to comprehend the spec it helps to read the many discussion threads rather than the spec itself. One relevant statement of the spec is:
> 
> > When using the discriminator, inline schemas will not be considered.

When using the discriminator, inline schemas will not be considered.

I read this as don't deserialize inline schemas when a discriminator exists in a composed schema.